src/osp/id_registry.h

```diff
@@ -0,0 +1,349 @@
+static_assert(std::is_same_v<underlying_int_type_t<int>, int>);
+static_assert(std::is_same_v<underlying_int_type_t<Test>, uint32_t>);
+
+//-----------------------------------------------------------------------------
+
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using id_int_t = underlying_int_type_t<TYPE_T>;
+    return TYPE_T(std::numeric_limits<id_int_t>::max());
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Generates reusable sequential IDs
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class IdRegistry
+{
+    using id_int_t = underlying_int_type_t<ID_T>;
+
+public:
+
+    IdRegistry() = default;
+    IdRegistry(size_t capacity) { reserve(capacity); };
+
+    /**
+     * @brief Create a single ID
+     *
+     * @return A newly created ID
+     */
+    [[nodiscard]] ID_T create()
+    {
+        ID_T output{ id_null<ID_T>() };
+        create(&output, 1);
+        return output;
+    }
+
+    /**
+     * @brief Create multiple IDs
+     *
+     * @param out   [out] Iterator out
+     * @param count [in] Number of IDs to generate
+     */
+    template<typename IT_T>
+    void create(IT_T out, size_t count);
+
+    /**
+     * @return Size required to fit all existing IDs, or allocated size if
+     *         reserved ahead of time
+     */
+    constexpr size_t capacity() const noexcept { return m_deleted.size(); }
+
+    /**
+     * @return Number of registered IDs
+     */
+    constexpr size_t size() const { return capacity() - m_deleted.count(); }
+
+    /**
+     * @brief Allocate space for at least n IDs
+     *
+     * @param n [in] Number of IDs to allocate for
+     */
+    void reserve(size_t n)
+    {
+        m_deleted.resize(n, true);
+    }
+
+    /**
+     * @brief Remove an ID. This will mark it for reuse
+     *
+     * @param id [in] ID to remove
+     */
+    void remove(ID_T id)
+    {
+        if ( ! exists(id))
+        {
+            throw std::runtime_error("ID does not exist");
+        }
+
+        m_deleted.set(id_int_t(id));
+    }
+
+    /**
+     * @brief Check if an ID exists
+     *
+     * @param id [in] ID to check
+     *
+     * @return true if ID exists
+     */
+    bool exists(ID_T id) const noexcept
+    {
+        return ! m_deleted.test(id_int_t(id));
+    };
+
+private:
+    HierarchicalBitset<uint64_t> m_deleted;
+
+}; // class IdRegistry
+
+template<typename ID_T, bool NO_AUTO_RESIZE>
+template<typename IT_T>
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::create(IT_T out, size_t count)
+{
+    if (m_deleted.count() < count)
+    {
+        // auto-resize
+        if constexpr (NO_AUTO_RESIZE)
+        {
+            throw std::runtime_error(
+                    "Reached max capacity with automatic resizing disabled");
+        }
+        else
+        {
+            reserve(std::max(capacity() + count, capacity() * 2));
+        }
+    }
+
+    m_deleted.take<IT_T, ID_T>(out, count);
```

Comment:
Are there any concerns about re-using deleted indices right away? One of the issues I recall facing in my first ECS engine was that you had to be careful recycling entity IDs since another entity's component may store that ID even though it's stale, and ends up referring to an entity that it shouldn't be.

Reply:
This is a minor concern, but still a valid one. If an entity is deleted, then ideally there shouldn't be any instances of it at all. It only makes an existing problem slightly worse, but the problem can't exist on its own.